"panic: bad uvarint: 0" during flatten()

Repro is with Zed commit 2d52400. We stumbled onto this issue while working on possible solutions for #4050.

```
$ zq -version
Version: v1.2.0-29-g2d52400f

$ zq -Z before.zson 
{
    value: {
        before: null (foo={bar:int32})
    }
}

$ zq -Z 'flatten(this)' before.zson
panic: bad uvarint: 0
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x65
github.com/brimdata/zed/runtime/op.(*Catcher).Pull.func1()
	/Users/phil/work/zed/runtime/op/catcher.go:25 +0x45
panic({0x16a32a0, 0xc0003d4810})
	/usr/local/opt/go/libexec/src/runtime/panic.go:884 +0x212
github.com/brimdata/zed/zcode.(*Iter).Next(0x2000000203000?)
	/Users/phil/work/zed/zcode/iter.go:26 +0x219
github.com/brimdata/zed/runtime/expr/function.(*Flatten).encode(0xc00004b200, {0xc0002d6a00?, 0x1, 0x16ea2c0?}, {0x193dc90?, 0xc0003ea330?}, {0xc0002d6aa0, 0x2, 0x2}, {0x0, ...})
	/Users/phil/work/zed/runtime/expr/function/flatten.go:89 +0x125
github.com/brimdata/zed/runtime/expr/function.(*Flatten).encode(0xc00004b200, {0xc0002d6a40?, 0x1, 0x10229a7?}, {0x193dc90?, 0xc0003ea330?}, {0xc0003d4800, 0x1, 0x1}, {0xc0002cf6d3, ...})
	/Users/phil/work/zed/runtime/expr/function/flatten.go:92 +0x2c8
github.com/brimdata/zed/runtime/expr/function.(*Flatten).encode(0xc00004b200, {0xc0002d6a80?, 0x1, 0x100b615?}, {0x193dc90?, 0xc0003ea330?}, {0xc00019f848, 0x0, 0x0}, {0xc0002cf6d2, ...})
	/Users/phil/work/zed/runtime/expr/function/flatten.go:92 +0x2c8
github.com/brimdata/zed/runtime/expr/function.(*Flatten).Call(0xc00004b200, {0x800e750, 0xc000013d40}, {0xc0002e5cb0, 0x1, 0x800e720?})
	/Users/phil/work/zed/runtime/expr/function/flatten.go:47 +0x230
github.com/brimdata/zed/runtime/expr.(*Call).Eval(0xc00007ed20, {0x800e720?, 0xc000013d40}, 0x8008648?)
	/Users/phil/work/zed/runtime/expr/eval.go:809 +0x84
github.com/brimdata/zed/runtime/op/yield.(*Proc).Pull(0xc0002e5ce0, 0x5?)
	/Users/phil/work/zed/runtime/op/yield/yield.go:31 +0x413
github.com/brimdata/zed/runtime/op.(*Single).Pull(0xc000013c98, 0x80)
	/Users/phil/work/zed/runtime/op/mux.go:134 +0x35
github.com/brimdata/zed/runtime/op.(*Catcher).Pull(0x19405c0?, 0x50?)
	/Users/phil/work/zed/runtime/op/catcher.go:28 +0x6d
github.com/brimdata/zed/runtime.(*Query).Pull(0x800e570?, 0x40?)
	/Users/phil/work/zed/runtime/query.go:92 +0x45
github.com/brimdata/zed/zbuf.CopyPuller({0x800e570, 0xc000013758}, {0x1938600, 0xc0002d6760})
	/Users/phil/work/zed/zbuf/batch.go:106 +0x53
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc00026a000, {0xc0000d2020, 0x2, 0x2})
	/Users/phil/work/zed/cli/zq/command.go:166 +0x798
github.com/brimdata/zed/pkg/charm.path.run({0xc0000c80a8?, 0x1, 0x1}, {0xc0000d2020?, 0x2, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x8b
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0xc0000061a0?, {0xc0000d2010, 0x3, 0x3})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x4a
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5d
```

It looks like it's choking on the named type, as I can't repro if it's a simple `null` or `null (string)`.

--- a/runtime/expr/function/flatten.go
+++ b/runtime/expr/function/flatten.go
@@ -1,8 +1,6 @@
 package function
 
 import (
-	"sort"
-
 	"github.com/brimdata/zed"
 	"github.com/brimdata/zed/pkg/field"
 	"github.com/brimdata/zed/zcode"
@@ -21,7 +19,6 @@ func NewFlatten(zctx *zed.Context) *Flatten {
 	return &Flatten{
 		entryTypes: make(map[zed.Type]zed.Type),
 		keyType:    zctx.LookupTypeArray(zed.TypeString),
-		mapper:     zed.NewMapper(zctx),
 		zctx:       zctx,
 	}
 }
@@ -32,55 +29,40 @@ func (n *Flatten) Call(ctx zed.Allocator, args []zed.Value) *zed.Value {
 	if typ == nil {
 		return &val
 	}
-	valType := n.mapper.Lookup(typ.ID())
-	if valType == nil {
-		types := n.collectTypes(typ.Columns)
-		types = dedupeTypes(types)
-		if len(types) == 1 {
-			valType = types[0]
-		} else {
-			valType = n.zctx.LookupTypeUnion(types)
-		}
-		n.mapper.EnterType(typ.ID(), valType)
-	}
+	inner := n.innerTypeOf(val.Bytes, typ.Columns)
 	n.Reset()
-	n.encode(typ.Columns, valType, field.Path{}, val.Bytes)
-	return ctx.NewValue(n.zctx.LookupTypeArray(valType), n.Bytes())
+	n.encode(typ.Columns, inner, field.Path{}, val.Bytes)
+	return ctx.NewValue(n.zctx.LookupTypeArray(inner), n.Bytes())
 }
 
-func (n *Flatten) collectTypes(cols []zed.Column) []zed.Type {
-	var types []zed.Type
+func (n *Flatten) innerTypeOf(b zcode.Bytes, cols []zed.Column) zed.Type {
+	types := n.appendTypes(nil, b, cols)
+	unique := zed.UniqueTypes(types)
+	if len(unique) == 1 {
+		return unique[0]
+	}
+	return n.zctx.LookupTypeUnion(unique)
+}
+
+func (n *Flatten) appendTypes(types []zed.Type, b zcode.Bytes, cols []zed.Column) []zed.Type {
+	it := b.Iter()
 	for _, col := range cols {
-		if typ := zed.TypeRecordOf(col.Type); typ != nil {
-			types = append(types, n.collectTypes(typ.Columns)...)
-		} else {
-			typ, ok := n.entryTypes[col.Type]
-			if !ok {
-				typ = n.zctx.MustLookupTypeRecord([]zed.Column{
-					zed.NewColumn("key", n.keyType),
-					zed.NewColumn("value", col.Type),
-				})
-				n.entryTypes[col.Type] = typ
-			}
-			types = append(types, typ)
+		val := it.Next()
+		if typ := zed.TypeRecordOf(col.Type); typ != nil && val != nil {
+			types = n.appendTypes(types, val, typ.Columns)
+			continue
 		}
+		typ, ok := n.entryTypes[col.Type]
+		if !ok {
+			typ = n.zctx.MustLookupTypeRecord([]zed.Column{
+				zed.NewColumn("key", n.keyType),
+				zed.NewColumn("value", col.Type),
+			})
+			n.entryTypes[col.Type] = typ
+		}
+		types = append(types, typ)
 	}
 	return types
-}
-
-func dedupeTypes(types []zed.Type) []zed.Type {
-	sort.SliceStable(types, func(i, j int) bool {
-		return zed.CompareTypes(types[i], types[j]) < 0
-	})
-	out := make([]zed.Type, 0, len(types))
-	var prev zed.Type
-	for _, typ := range types {
-		if typ != prev {
-			out = append(out, typ)
-			prev = typ
-		}
-	}
-	return out
 }
 
 func (n *Flatten) encode(cols []zed.Column, inner zed.Type, base field.Path, b zcode.Bytes) {
@@ -88,7 +70,7 @@ func (n *Flatten) encode(cols []zed.Column, inner zed.Type, base field.Path, b z
 	for _, col := range cols {
 		val := it.Next()
 		key := append(base, col.Name)
-		if typ := zed.TypeRecordOf(col.Type); typ != nil {
+		if typ := zed.TypeRecordOf(col.Type); typ != nil && val != nil {
 			n.encode(typ.Columns, inner, key, val)
 			continue
 		}

--- a/runtime/expr/function/ztests/flatten.yaml
+++ b/runtime/expr/function/ztests/flatten.yaml
@@ -5,9 +5,13 @@ input: |
   {a:1,b:null}
   {a:1,b:{c:"foo",d:[1,2,3]}}
   {a:1,b:{c:2,d:{e:4,f:5}}}
+  {foo:null({bar:int32})}
+  {foo:{bar:2(int32)}}
 
 output: |
   127.0.0.1
   [{key:["a"],value:1},{key:["b"],value:null}]
   [{key:["a"],value:1},{key:["b","c"],value:"foo"},{key:["b","d"],value:[1,2,3]}]
   [{key:["a"],value:1},{key:["b","c"],value:2},{key:["b","d","e"],value:4},{key:["b","d","f"],value:5}]
+  [{key:["foo"],value:null({bar:int32})}]
+  [{key:["foo","bar"],value:2(int32)}]

